### PR TITLE
feat(api)!: raise DecodingError on derived-read + lister drift (#1344)

### PR DESCRIPTION
**Pull request by teng-lin**
## What

Three RPC "derived-read / lister" sites now raise `DecodingError` on a
*structurally-unrecognized* payload shape instead of silently collapsing to
`False` / `[]`, so callers can distinguish a genuine miss from server-side
schema drift (#1344):

- **`sources.check_freshness()`** — drift raises; recognized shapes keep their
  bool (`True` / `[]` / `[[null, true, …]]` → fresh,
  `False` / `[[null, false, …]]` → stale).
- **the note lister** (`NoteService._extract_note_row_container`) — a *truthy
  non-list* payload raises; `None` / `[]` still mean an empty notebook (`[]`).
- **the artifact raw lister** (`ArtifactListingService.list_raw`) — same: a
  truthy non-list raises, a falsy payload returns `[]`.

The uniform-empty derived-read contract (`get_summary` / `get_description` /
`get_guide` → empty, `get_tree` → `None`) and the `research.poll` NOT_FOUND
sentinel are unchanged.

## Why

A malformed payload that silently becomes empty hides schema drift behind a
"not found". Raising lets `notes` / `mind_maps` `get()`/`get_or_none()` and the
freshness check tell a real miss from drift — the consistent fail-loud contract
ADR-0019 is converging the SDK toward.

## Notes

- The freshness shape-interpretation is extracted into a new module-level helper
  `interpret_source_freshness()` in `_row_adapters/sources.py` — the right layer
  for positional-RPC decoding — which also keeps the over-budget `_sources.py`
  under its module-size ratchet (ceiling lowered 1023 → 1015).
- The mind-map sub-fetch in `list_artifacts` still degrades gracefully on a
  *transient* outage (`RPCError` / `HTTPError`), but now re-raises `DecodingError`
  so drift in the secondary fetch is surfaced rather than masked as "no mind
  maps" — matching the swallow's own "temporarily unavailable" intent.
- Deliberate clean break (no value-level warning is possible for "this future
  payload shape will be rejected"), tracked as an exempted entry in the v0.8.0
  breaking-changes registry. Return types are unchanged, so the public-API compat
  audit needs no allowlist entry.

## Tests

- Direct unit coverage for `interpret_source_freshness` (recognized → bool,
  drift → raise), the note/artifact listers (truthy non-list → raise), and
  `list_artifacts` (mind-map drift propagates; transient outage still degrades).
- `check_freshness` integration tests flip the drift shapes to expect
  `DecodingError` while keeping recognized-stale → `False`.
- Full suite green; mypy + ruff clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Stricter handling of malformed API responses for artifact listings, notes retrieval, and source-freshness checks: unexpected payload shapes now surface explicit decoding errors instead of being treated as empty results.

* **Tests**
  * Expanded unit and integration tests to cover stricter decoding behavior, freshness interpretation, and artifact-listing edge cases.

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/_row_adapters/sources.py [hunk 1]
```diff
@@ -525,3 +526,32 @@ def status(self) -> SourceStatus:
...
+    if result is True:
+        return True
+    if result is False:
+        return False
+    if isinstance(result, list):
+        if len(result) == 0:
+            return True  # empty array = fresh
+        first = result[0]
+        if isinstance(first, list) and len(first) > 1:
+            return first[1] is True
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_row_adapters/sources.py]: Addressed in 69868fa7: `interpret_source_freshness` now raises `DecodingError` on a non-boolean freshness flag — only a clean `True`/`False` at `first[1]` is a recognized shape; any other value falls through to the raise instead of being coerced to stale.

File: tests/unit/test_row_adapters.py [hunk 2]
```diff
@@ -1702,3 +1706,22 @@ def test_does_not_mutate_wrapped_row(self) -> None:
...
+    @pytest.mark.parametrize("payload", [False, [[None, False, ["src"]]]])
+    def test_stale_shapes_return_false(self, payload: object) -> None:
+        assert interpret_source_freshness(payload) is False
+
+    @pytest.mark.parametrize("payload", [None, "some_value", ["some_value"], [[None]]])
+    def test_unrecognized_shapes_raise(self, payload: object) -> None:
+        # None, a bare scalar, a list whose first element is a non-list scalar,
+        # and a too-short nested list are all drift -> raise, not a silent bool.
+        with pytest.raises(DecodingError):
+            interpret_source_freshness(payload)
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/unit/test_row_adapters.py]: Addressed in 69868fa7: added `[[None, "unexpected", [...]]]`, `[[None, None, [...]]]`, and `[[None, 1, [...]]]` to `test_unrecognized_shapes_raise`.

File: src/notebooklm/_note_service.py [hunk 3]
```diff
@@ -124,8 +124,16 @@ def _extract_note_row_container(self, result: Any) -> list[Any]:
...
-        if not result or not isinstance(result, list):
+        if not result:
             return []
+        if not isinstance(result, list):
+            # A truthy non-list payload is schema drift, not a legitimately empty
+            # notebook — raise so notes/mind_maps get()/get_or_none can tell a
+            # miss from drift instead of silently collapsing to ``[]``.
+            raise DecodingError(
+                f"Unrecognized GET_NOTES_AND_MIND_MAPS payload shape: {result!r}",
+                method_id=RPCMethod.GET_NOTES_AND_MIND_MAPS.value,
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_note_service.py]: Addressed in 69868fa7: the drifted payload now goes through `raw_response=repr(result)` (truncated + scrubbed by `_truncate_response_preview`) with a generic message, across all three raise sites (`_note_service`, `_artifact/listing`, and `interpret_source_freshness`), so notebook content is no longer spliced into the exception message.

File: src/notebooklm/_row_adapters/sources.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_row_adapters/sources.py]: Addressed in 69868fa7: a non-boolean freshness flag now raises `DecodingError` instead of coercing to stale — only `True`/`False` are recognized.

File: tests/unit/test_row_adapters.py [hunk 4]
```diff
@@ -1702,3 +1706,22 @@ def test_does_not_mutate_wrapped_row(self) -> None:
...
+
+    @pytest.mark.parametrize("payload", [True, [], [[None, True, ["src"]]]])
+    def test_fresh_shapes_return_true(self, payload: object) -> None:
+        assert interpret_source_freshness(payload) is True
+
+    @pytest.mark.parametrize("payload", [False, [[None, False, ["src"]]]])
+    def test_stale_shapes_return_false(self, payload: object) -> None:
+        assert interpret_source_freshness(payload) is False
+
+    @pytest.mark.parametrize("payload", [None, "some_value", ["some_value"], [[None]]])
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/unit/test_row_adapters.py]: Addressed in 69868fa7: added non-boolean freshness-flag cases to the drift parametrization in `test_row_adapters.py`.
